src/ergw_tei_mngr.erl

```diff
@@ -0,0 +1,168 @@
+release_tei(Port, TEI) ->
+    with_keyfun(Port, release_tei(_, TEI, _)).
+
+%% release_tei/3
+release_tei(Name, TEI, KeyFun) ->
+    gen_server:call(?SERVER, {release_tei, Name, TEI, KeyFun}).
+
+%%%===================================================================
+%%% Options Validation
+%%%===================================================================
+
+validate_option({Prefix, Len} = Value)
+  when is_integer(Prefix), is_integer(Len),
+       Len >= 0, Len =< 8, Prefix >= 0 ->
+    case ?POW(2, Len) of
+	X when X > Prefix -> Value;
+	_ ->
+	    throw({error, {options, {teid, Value}}})
+    end;
+validate_option(Value) ->
+    throw({error, {options, {teid, Value}}}).
+
+%%%===================================================================
+%%% gen_statem callbacks
+%%%===================================================================
+
+callback_mode() -> handle_event_function.
+
+init([]) ->
+    {ok, {Prefix, PrefixLen}} = application:get_env(ergw, teid),
+    TID = ets:new(?SERVER, [ordered_set, {keypos, 1}]),
+    NodeId = node(),
+    BlockId = iolist_to_binary(["teid-prefix-", integer_to_list(Prefix)]),
+
+    Data = #data{node_id = NodeId,
+		  block_id = BlockId,
+		  prefix = Prefix,
+		  prefix_len = PrefixLen,
+		  tid = TID,
+		  save_cnt = ?SAVE_COUNTER},
+
+    {ok, #{}, Data}.
+
+handle_event(state_timeout, save, State, Data) ->
+    save_data(State, Data),
+    {keep_state, Data#data{save_cnt = ?SAVE_COUNTER}};
+
+handle_event({call, From}, {alloc_tei, Name, KeyFun}, State, Data) ->
+    RndState = maybe_init_rnd(Name, State),
+    alloc_tei(From, Name, KeyFun, ?MAX_TRIES, RndState, State, Data);
+
+handle_event({call, From}, {release_tei, _Name, TEI, KeyFun}, _State, #data{tid = TID}) ->
+    ets:delete(TID, KeyFun(TEI)),
+    Actions = [{reply, From, ok}, state_timeout()],
+    {keep_state_and_data, Actions};
+
+handle_event({call, From}, Request, _State, _Data) ->
+    Reply = {error, {badarg, Request}},
+    {keep_state_and_data, [{reply, From, Reply}]};
+
+handle_event(_, _Msg, _State, _Data) ->
+    keep_state_and_data.
+
+terminate(_Reason, _State, _Data) ->
+    ok.
+
+code_change(_OldVsn, State, Data, _Extra) ->
+    {ok, State, Data}.
+
+%%%=========================================================================
+%%%  internal functions
+%%%=========================================================================
+
+maybe_init_rnd(Name, State) when is_map_key(Name, State) ->
+    maps:get(Name, State);
+maybe_init_rnd(_, _) ->
+    rand:seed_s(exrop).
+
+state_timeout() ->
+    {state_timeout, ?SAVE_TIMEOUT, save}.
+
+serialize_data(State, #data{tid = TID}) ->
+    SeedL = maps:fold(fun(K,V,L) -> [{K, rand:export_seed_s(V)}|L] end, [], State),
+    Term =
+	#{seed  => SeedL,
+	  teids => ets:tab2list(TID)},
+    term_to_binary(Term, [{minor_version, 2}, compressed]).
+
+save_data(State, #data{node_id = _NodeId, block_id = _BlockId} = Data) ->
+    _Serialized = serialize_data(State, Data).
```

Comment:
state_timeout, serialize_data and save_data are not really used in this version.